Added 3 fuzzers

AdamKorcz opened the pull request:
This PR answers https://github.com/gravitational/teleport/issues/2927 by adding 3 fuzzers to Teleport.

The fuzzers add no extra functionality to Teleport. They are tests themselves that run their targets with lots of pseudo-random data.

I suggest integrating Teleport into OSS-fuzz to allow the fuzzers to run continuously. I can confirm that these 3 fuzzers run on OSS-fuzz's infrastructure, and integration would mean that the fuzzers would run a couple of times per week. If a bug is found during those runs, detailed reports will be sent to the maintainers. OSS-fuzz is a free service that is offered with an implied expectation that bugs are fixed, so that the resources spent on fuzzing Teleport are put to good use.

I will be happy to integrate Teleport into OSS-fuzz. All I need are the email addresses for the maintainers to which potential bug reports should be sent. These email addresses will be publicly available and can be changed anytime.

I will be happy to work more on fuzzing Teleport in the future.

AdamKorcz committed:
Added 3 fuzzers

gravitational-jenkins commented:
Can one of the admins verify this patch?

awly commented:
ok to test

awly reviewed (commented):
Thanks for the PR!
I agree that we could use some fuzzing. Will double-check with the team, in case anyone has objections.

In the meantime, could you move the `Fuzz` functions to standalone packages? Perhaps in subdirectories like `lib/fuzz/proxysubsysrequest`, `lib/fuzz/proxyjump` etc

Also, add a README in `lib/fuzz`, linking to OSS-fuzz and https://github.com/dvyukov/go-fuzz

webvictim reviewed (changes requested):
As @awly said, thanks for the PR. We'd appreciate it if you could address his feedback.

AdamKorcz commented:
Hi. Thank you for the feedback. I haven't forgotten about this or am ignoring it. I will get to it as soon as I can.

webvictim commented:
No problem! Just marking that it needs changes so it’s out of our review queue for now :)

AdamKorcz committed:
Moved fuzzers to a standalone package and removed a fuzzer for an unexported target

AdamKorcz commented:
I have moved the fuzzers to a standalone package and have removed the fuzzer for the unexported target for now. 

Furthermore I have added a README with instructions on running the fuzzers locally.

awly commented on lib/fuzz/README.md:
```diff
@@ -0,0 +1,13 @@
+All fuzzers are stored in fuzz.go and are implemented using [go-fuzz](https://github.com/dvyukov/go-fuzz)
+
+To run a fuzzer locally, follow these steps:
+1. go get github.com/google/zoekt
```
this is not needed

AdamKorcz replied:
This has been updated.

awly commented on lib/fuzz/fuzz.go:
```diff
@@ -0,0 +1,22 @@
+	"github.com/gravitational/teleport/lib/utils/parse"
+)
+
+func Fuzzutils(data []byte) int {
```
`go-fuzz` requires the function to be named `Fuzz` exactly, without a suffix. This means that there's only 1 fuzzer per package/directory. These individual fuzzers must live in sub-directories of `lib/fuzz`.

If you want, we can merge this in and I'll move things around afterwards as needed.

AdamKorcz replied:
I would like to mention that the name of each harness is flexible if you opt to fuzz continuously through oss-fuzz. As a reference, check out [Prometheus' fuzzers](https://github.com/prometheus/prometheus/blob/master/promql/fuzz.go). These run fine through oss-fuzz. 
As mentioned earlier, I have the setup ready to integrate teleport into oss-fuzz, if you would like to proceed in that direction.
I am also fine with merging now and moving things afterwards.

awly replied:
Ahh, I see. Please update `README.md` to point to `oss-fuzz` instead of `go-fuzz`, because you can't actually run these fuzzers locally with `go-fuzz`.

P.S. sorry for the response delay!

benarent commented:
@AdamKorcz PTAL at the code review.

AdamKorcz commented:
@benarent Thanks for the ping. I will get to it this week!

AdamKorcz committed:
Minor update to url

AdamKorcz committed:
Minor update to wrong url

awly commented on lib/fuzz/fuzz.go:
```diff
@@ -0,0 +1,22 @@
+	"github.com/gravitational/teleport/lib/utils/parse"
+)
+
+func Fuzzutils(data []byte) int {
```
```suggestion
func FuzzParseProxyJump(data []byte) int {
```

awly commented on lib/fuzz/fuzz.go:
```diff
@@ -0,0 +1,22 @@
+	return 1
+}
+
+func Fuzzparse(data []byte) int {
```
```suggestion
func FuzzParseRoleVariable(data []byte) int {
```

awly reviewed (approved):
@AdamKorcz sorry for the response delay!
LGTM.
Please rebase the PR and address the last few comments, I'll merge this PR afterwards.

AdamKorcz committed:
Updated fuzzer names and modified fuzzer README

AdamKorcz commented:
I just noticed that `parse.RoleVariable` has changed to `NewExpression`. I will be updating this shortly.

AdamKorcz committed:
Updated function name in fuzzer

AdamKorcz committed:
go-fmt'd fuzz.go and added build instruction

Commit:
Merge branch 'master' into master

awly commented:
Sorry for the long turnaround. LGTM.
@webvictim need your approval to unblock the merge.

webvictim committed:
Merge branch 'master' into master

Commit:
Merge branch 'master' into master
